File: 121. Best Time to Buy and Sell Stock/121.py
class Solution:
    def maxProfit(self, prices) -> int:
        # if len(prices)<2:
        #     return 0
        # max = prices[1:]
        # max.sort()
        # max = max[-1]
        # maxmax = 0
        # for index,i in enumerate(prices):
        #     if i != max:
        #         if max-i>maxmax:
        #             maxmax = max-i
        #     else:
        #         if index<len(prices)-1:
        #             max = prices[index+1:]
        #             max.sort()
        #             max = max[-1]
        # return maxmax
        if len(prices) == 0:
            return 0
        min = prices[0]
        max = 0
        for i in prices:
            if min>=i:
                min = i
            else:
                if i-min>max:
                    max=i-min
        return max

a = Solution()
print(a.maxProfit([7,1,5,3,6,4]))

File: 121. Best Time to Buy and Sell Stock/121.go
package main

import "fmt"

func main(){
	fmt.Println(maxProfit([]int{7,6,4,3,1}))
}

func maxProfit(prices []int) int {
    if len(prices) == 0{
		return 0
	}
	minprice := prices[0]
	maxvalue := 0

	for _,value := range prices{
		if minprice>value{
			minprice = value
		}else{
			if value-minprice > maxvalue{
				maxvalue = value-minprice
			}
		}
	}
	return maxvalue
}

